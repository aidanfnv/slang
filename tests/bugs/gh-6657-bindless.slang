//TEST:REFLECTION:-stage compute -entry computeMain -target cuda -no-codegen
//TEST:REFLECTION:-stage compute -entry computeMain -target metallib -no-codegen

// Test for bindless targets, DescriptorHandle should decompose to uint64_t handle pointers
// This complements gh-6657.slang which tests SPIR-V decomposition to uint2

struct BindlessData
{
    float4 data1;
    DescriptorHandle<Texture2D> textureHandle;
    DescriptorHandle<RWStructuredBuffer<float4>> bufferHandle;
    float4 data2;
}

[numthreads(1,1,1)]
void computeMain(BindlessData bindlessStruct)
{
    // Simple usage to avoid dead code elimination
}
