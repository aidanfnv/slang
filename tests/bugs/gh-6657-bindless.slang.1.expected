result code = 0
standard error = {
}
standard output = {
{
    "parameters": [

    ],
    "entryPoints": [
        {
            "name": "computeMain",
            "stage": "compute",
            "parameters": [
                {
                    "name": "bindlessStruct",
                    "bindings": [
                        {"kind": "shaderResource", "index": 0},
                        {"kind": "constantBuffer", "index": 1},
                        {"kind": "uniform", "offset": 0, "size": 32, "elementStride": 0}
                    ],
                    "type": {
                        "kind": "struct",
                        "name": "BindlessData",
                        "fields": [
                            {
                                "name": "data1",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 4,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 0, "size": 16, "elementStride": 4}
                            },
                            {
                                "name": "textureHandle",
                                "type": {
                                    "kind": "resource",
                                    "baseShape": "texture2D",
                                    "resultType": {
                                        "kind": "vector",
                                        "elementCount": 4,
                                        "elementType": {
                                            "kind": "scalar",
                                            "scalarType": "float32"
                                        }
                                    }
                                },
                                "binding": {"kind": "shaderResource", "index": 0}
                            },
                            {
                                "name": "bufferHandle",
                                "type": {
                                    "kind": "resource",
                                    "baseShape": "structuredBuffer",
                                    "access": "readWrite",
                                    "resultType": {
                                        "kind": "vector",
                                        "elementCount": 4,
                                        "elementType": {
                                            "kind": "scalar",
                                            "scalarType": "float32"
                                        }
                                    }
                                },
                                "binding": {"kind": "constantBuffer", "index": 0}
                            },
                            {
                                "name": "data2",
                                "type": {
                                    "kind": "vector",
                                    "elementCount": 4,
                                    "elementType": {
                                        "kind": "scalar",
                                        "scalarType": "float32"
                                    }
                                },
                                "binding": {"kind": "uniform", "offset": 16, "size": 16, "elementStride": 4}
                            }
                        ]
                    }
                }
            ],
            "threadGroupSize": [1, 1, 1]
        }
    ]
}
}
